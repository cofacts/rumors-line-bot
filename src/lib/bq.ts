/**
 * BigQuery client and schema
 */

import { BigQuery } from '@google-cloud/bigquery';

const bqDataset = new BigQuery().dataset(
  process.env.BIGQUERY_ANALYTICS_DATASET || ''
);

export function insertEventBatch(eventBatch: EventBatch) {
  const table = bqDataset.table(TABLE);
  return table.insert(eventBatch, { schema: bqSchema });
}
