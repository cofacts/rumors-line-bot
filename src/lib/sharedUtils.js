/**
 * Constants and functions shared across LIFF, webhook and other parts
 */

import { t } from 'ttag';
import dateFnsFormat from 'date-fns/format';
import dateFnsFormatDistanceToNow from 'date-fns/formatDistanceToNow';
import GraphemeSplitter from 'grapheme-splitter';

const SITE_URLS = (process.env.SITE_URLS || 'https://cofacts.tw').split(',');

/**
 * prefilled text for LIFF sendMessage()
 */
export const VIEW_ARTICLE_PREFIX = `ðŸ“ƒ ${t`See replies of`}:\n`;

/**
 * @param {string} articleId
 * @returns {string} The article's full URL
 */
export function getArticleURL(articleId) {
  return `${SITE_URLS[0]}/article/${articleId}`;
}

/**
 * Extracts Article ID from message with just article URL, or prefix + article URL.
 *
 * @param {string} message
 * @returns {string} Extracts article ID from message, or empty string if the given string is not article URL.
 */
export function extractArticleId(message) {
  if (message.startsWith(VIEW_ARTICLE_PREFIX)) {
    message = message.replace(VIEW_ARTICLE_PREFIX, '');
  }

  for (const siteUrl of SITE_URLS) {
    const articleUrlPrefix = `${siteUrl}/article/`;
    if (message.startsWith(articleUrlPrefix)) {
      return message.replace(articleUrlPrefix, '');
    }
  }

  return '';
}

let locale = require(`date-fns/locale/${(process.env.LOCALE || 'en_US').replace(
  '_',
  '-'
)}`);

// Svelte fix: babel interprets default automatically, but svelte doesn't.
/* istanbul ignore next */
locale = locale.default ? locale.default : locale;

function formatAbsolute(date, format = 'PP', config = {}) {
  return dateFnsFormat(date, format, { ...config, locale });
}

export function formatDistanceToNow(date, config = {}) {
  return dateFnsFormatDistanceToNow(date, { ...config, locale });
}

const THRESHOLD = 86400 * 2 * 1000; // 2 days in milliseconds

export function format(date) {
  const now = new Date();

  if (now - date < THRESHOLD) {
    const dateStr = formatDistanceToNow(date);
    return t`${dateStr} ago`;
  }

  return formatAbsolute(date);
}

export function createTypeWords(type) {
  switch (type) {
    case 'RUMOR':
      return t`Contains misinformation`;
    case 'NOT_RUMOR':
      return t`Contains true information`;
    case 'OPINIONATED':
      return t`Contains personal perspective`;
    case 'NOT_ARTICLE':
      return t`Invalid request`;
  }
  return 'Undefined';
}

const splitter = new GraphemeSplitter();

// When document title is too long, event will be dropped.
// See: https://github.com/cofacts/rumors-line-bot/issues/97
//
const DOCUMENT_TITLE_LENGTH = 800;

/**
 * @param {string} title
 * @returns {string} valid title for Google Analytics
 */
export function gaTitle(title) {
  return splitter
    .splitGraphemes(title)
    .slice(0, DOCUMENT_TITLE_LENGTH)
    .join('');
}
