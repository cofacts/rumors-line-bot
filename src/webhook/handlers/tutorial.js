import { t } from 'ttag';
import { ellipsis, createPostbackAction, createReplyMessages } from './utils';
import ga from 'src/lib/ga';

/**
 * When updating tutorial images(static/img/), change TUTORIAL_IMAGE_VERSION for cache busting
 */
const TUTORIAL_IMAGE_VERSION = '20201105';

/**
 * Fixed inputs that indicate which reply should `tutorial` function return
 */
export const TUTORIAL_STEPS = {
  // From rich menu
  RICH_MENU: `📖 ${t`tutorial`}`,
  // From flex message button
  SIMULATE_FORWARDING_MESSAGE: t`Simulates forwarding a message`,
  // From quick reply, Note: it should be less than 20 charactors
  PROVIDE_PERMISSION_SETUP: `💡 ${t`Cool, I got it!`}`,
  EXPLAN_CHATBOT_FLOW_AND_PROVIDE_PERMISSION_SETUP: `🤔 ${t`What happened?`}`,
  PROVIDE_PERMISSION_SETUP_WITH_EXPLANATION: `❓ ${t`Why`}`,
  SETUP_DONE: `👌 ${t`Done!`}`,
  SETUP_LATER: `⏱️ ${t`Later`}`,
};

/**
 * @param {string} imageUrl
 * @param {string} text
 * @returns {object} Flex message contents object
 */
function createImageTextBubble(imageUrl, text) {
  return {
    type: 'bubble',
    body: {
      type: 'box',
      layout: 'vertical',
      contents: [
        {
          type: 'image',
          url: imageUrl,
          size: 'full',
          aspectMode: 'cover',
          aspectRatio: '1:1',
          gravity: 'top',
        },
        {
          type: 'box',
          layout: 'vertical',
          contents: [
            {
              type: 'text',
              text: text,
              size: 'lg',
              weight: 'bold',
              wrap: true,
            },
          ],
          position: 'relative',
          offsetBottom: '0px',
          offsetStart: '0px',
          offsetEnd: '0px',
          paddingAll: '20px',
          paddingTop: '18px',
          flex: 1,
          justifyContent: 'center',
        },
      ],
      paddingAll: '0px',
    },
  };
}

/**
 * @param {string} label Act as quickReply's label and postback's input and displayText
 * @param {string} sessionId Search session ID
 * @param {string} postbackState Used by `handleInput` to determine which handler to call
 * @returns {object} quickReply items object
 */
function createQuickReplyPostbackItem(label, sessionId, postbackState) {
  return {
    type: 'action',
    action: createPostbackAction(label, label, label, sessionId, postbackState),
  };
}

/**
 * @returns {object} Flex message object
 */
export function createGreetingMessage() {
  const text = t`This is a chatbot that looks up suspicious forwarded messages for you. Here is how to use me:`;
  const imageUrl = `${
    process.env.RUMORS_LINE_BOT_URL
  }/static/img/greeting.png?cachebust=${TUTORIAL_IMAGE_VERSION}`;

  return {
    type: 'flex',
    altText: ellipsis(text, 300),
    contents: {
      type: 'carousel',
      contents: [createImageTextBubble(imageUrl, text)],
    },
  };
}

/**
 * @param {string} sessionId Search session ID
 * @returns {object} Flex message object
 */
export function createTutorialMessage(sessionId) {
  const textStep1 = `1. ${t`When receiving a message from elsewhere`}`;
  const textStep2 = `2. ${t`Long press and share`}`;
  const textStep3 = `3. ${t`Select Cofacts to share`}`;
  const textStep4 = `4. ${t`Cofacts replies with a crowd-sourced fact-check or chatbot replies`}`;

  const imageUrlStep1 = `${
    process.env.RUMORS_LINE_BOT_URL
  }/static/img/tutorial1.png?cachebust=${TUTORIAL_IMAGE_VERSION}`;
  const imageUrlStep2 = `${
    process.env.RUMORS_LINE_BOT_URL
  }/static/img/tutorial2.png?cachebust=${TUTORIAL_IMAGE_VERSION}`;
  const imageUrlStep3 = `${
    process.env.RUMORS_LINE_BOT_URL
  }/static/img/tutorial3.png?cachebust=${TUTORIAL_IMAGE_VERSION}`;
  const imageUrlStep4 = `${
    process.env.RUMORS_LINE_BOT_URL
  }/static/img/tutorial4.png?cachebust=${TUTORIAL_IMAGE_VERSION}`;

  const askForForwardingMessage = t`Wanna try it out? Just forward a message to me!`;
  const buttonLabel = TUTORIAL_STEPS['SIMULATE_FORWARDING_MESSAGE'];
  const displayText =
    '只要每天早上一杯地瓜葉牛奶。不僅有效降低三高，甚至連痛風也沒了；此外，地瓜葉牛奶的作法也很簡單，只要先將地瓜葉川燙過後，再加入鮮奶打成汁即可。';

  return {
    type: 'flex',
    altText: ellipsis(
      textStep1 + '\n' + textStep2 + '\n' + textStep3 + '\n' + textStep4,
      300
    ),
    contents: {
      type: 'carousel',
      contents: [
        createImageTextBubble(imageUrlStep1, textStep1),
        createImageTextBubble(imageUrlStep2, textStep2),
        createImageTextBubble(imageUrlStep3, textStep3),
        createImageTextBubble(imageUrlStep4, textStep4),
        {
          type: 'bubble',
          body: {
            type: 'box',
            layout: 'vertical',
            contents: [
              {
                type: 'text',
                text: askForForwardingMessage,
                size: 'lg',
                weight: 'bold',
                wrap: true,
              },
            ],
          },
          footer: {
            type: 'box',
            layout: 'horizontal',
            contents: [
              {
                type: 'button',
                action: createPostbackAction(
                  buttonLabel,
                  TUTORIAL_STEPS['SIMULATE_FORWARDING_MESSAGE'],
                  displayText,
                  sessionId,
                  'TUTORIAL'
                ),
                style: 'primary',
                color: '#ffb600',
              },
            ],
          },
        },
      ],
    },
  };
}

/**
 * @returns {object} Flex message object
 */
function createEndingMessage() {
  const text = `${t`This is the end of the tutorial. Next time when you receive a suspicious message, don't hesitate to forward it to me!`} 🤗`;
  const imageUrl = `${
    process.env.RUMORS_LINE_BOT_URL
  }/static/img/endoftutorial.png?cachebust=${TUTORIAL_IMAGE_VERSION}`;

  return {
    type: 'flex',
    altText: ellipsis(text, 300),
    contents: {
      type: 'carousel',
      contents: [createImageTextBubble(imageUrl, text)],
    },
  };
}

/**
 * @param {string} sessionId Search session ID
 * @returns {object[]} message objects
 */
function createMockReplyMessages(sessionId) {
  const reply = {
    type: 'RUMOR',
    reference:
      'http://www.mygopen.com/2017/06/blog-post_26.html\n神奇的地瓜葉？搭配鮮奶遠離三高？謠言讓醫生說：有痛風或是腎臟不好的人要小心！',
    text:
      '基本上地瓜葉其實單吃就有效果，牛奶、豆漿可加可不加，民眾不用迷信。 三高或是糖尿病的患者還是要搭配醫生的治療，不能單靠吃地瓜葉就想將身體調養好，民眾千萬要注意。\n另外地瓜葉內還有鉀和鈉，對於有痛風或是腎臟不好的民眾反而會造成負擔，因此並不建議食用。',
  };
  const article = { replyCount: 1 };
  const selectedArticleId = '2sn80q5l5mzi0';

  const replies = createReplyMessages(reply, article, selectedArticleId);
  // put quickreply into last message
  replies[replies.length - 1]['quickReply'] = {
    items: [
      createQuickReplyPostbackItem(
        TUTORIAL_STEPS['PROVIDE_PERMISSION_SETUP'],
        sessionId,
        'TUTORIAL'
      ),
      createQuickReplyPostbackItem(
        TUTORIAL_STEPS['EXPLAN_CHATBOT_FLOW_AND_PROVIDE_PERMISSION_SETUP'],
        sessionId,
        'TUTORIAL'
      ),
    ],
  };
  return replies;
}

/**
 * @param {string} message
 * @returns {object} Flex message object
 */
function createPermissionSetupDialog(message) {
  const buttonLabel = t`Setup permission`;
  const buttonUri = `${
    process.env.LIFF_URL
  }?p=setting&utm_source=rumors-line-bot&utm_medium=tutorial`;

  return {
    type: 'flex',
    altText: ellipsis(message, 300),
    contents: {
      type: 'bubble',
      direction: 'ltr',
      body: {
        type: 'box',
        layout: 'vertical',
        spacing: 'none',
        margin: 'none',
        contents: [
          {
            type: 'text',
            contents: [
              {
                type: 'span',
                text: message,
              },
            ],
            flex: 0,
            gravity: 'top',
            weight: 'regular',
            wrap: true,
          },
        ],
      },
      footer: {
        type: 'box',
        layout: 'horizontal',
        contents: [
          {
            type: 'button',
            action: {
              type: 'uri',
              label: buttonLabel,
              uri: buttonUri,
            },
            style: 'primary',
            color: '#ffb600',
          },
        ],
      },
    },
  };
}

export default function tutorial(params) {
  let { data, event, issuedAt, replies, userId, isSkipUser } = params;

  const replyProvidePermissionSetup = `${t`You are smart`} 😊`;
  const replySetupLater = t`OK. When we ask for feedback from you, the permission dialog will pop-up again.`;

  const askForPermissionSetup = t`To wrap up, please finish your permission settings so that I can provide a smoother experience.`;
  const explanPersmissionSetup = t`When I provide hoax-busting replies to you, I would like to ask you for any feedback on the crowd-sourced reply.
  In order to achieve that, I need to ask for your permission to "Send messages to chats".
  The permission will be used to send only this one message of yours back to this particular chatroom.
  You can still use Cofacts without granting me this permission. When we ask for feedback from you, the permission dialog will pop-up again.`;
  const explanChatbotFlow =
    `⬆️ ${t`What you see above is a simulation of what you may see after you forward a message to Cofacts.`}\n\n` +
    `📚 ${t`Cofacts has a database of hoax messages and replies.`}\n\n` +
    `📲 ${t`When you send a message to me, I look up the message in our database and return the results I found.`}\n\n` +
    `🆕 ${t`If I can't find anything, I will ask you about sending your message to that database.`}`;
  if (!process.env.RUMORS_LINE_BOT_URL) {
    throw new Error('RUMORS_LINE_BOT_URL undefined');
  } else if (event.input === TUTORIAL_STEPS['RICH_MENU']) {
    replies = [createTutorialMessage(data.sessionId)];
  } else if (event.input === TUTORIAL_STEPS['SIMULATE_FORWARDING_MESSAGE']) {
    replies = createMockReplyMessages(data.sessionId);
  } else if (event.input === TUTORIAL_STEPS['PROVIDE_PERMISSION_SETUP']) {
    replies = [
      {
        type: 'text',
        text: replyProvidePermissionSetup,
      },
      {
        ...createPermissionSetupDialog(askForPermissionSetup),
        quickReply: {
          items: [
            createQuickReplyPostbackItem(
              TUTORIAL_STEPS['SETUP_DONE'],
              data.sessionId,
              'TUTORIAL'
            ),
            createQuickReplyPostbackItem(
              TUTORIAL_STEPS['SETUP_LATER'],
              data.sessionId,
              'TUTORIAL'
            ),
            createQuickReplyPostbackItem(
              TUTORIAL_STEPS['PROVIDE_PERMISSION_SETUP_WITH_EXPLANATION'],
              data.sessionId,
              'TUTORIAL'
            ),
          ],
        },
      },
    ];
  } else if (
    event.input ===
    TUTORIAL_STEPS['EXPLAN_CHATBOT_FLOW_AND_PROVIDE_PERMISSION_SETUP']
  ) {
    replies = [
      {
        type: 'text',
        text: explanChatbotFlow,
      },
      {
        ...createPermissionSetupDialog(askForPermissionSetup),
        quickReply: {
          items: [
            createQuickReplyPostbackItem(
              TUTORIAL_STEPS['SETUP_DONE'],
              data.sessionId,
              'TUTORIAL'
            ),
            createQuickReplyPostbackItem(
              TUTORIAL_STEPS['SETUP_LATER'],
              data.sessionId,
              'TUTORIAL'
            ),
            createQuickReplyPostbackItem(
              TUTORIAL_STEPS['PROVIDE_PERMISSION_SETUP_WITH_EXPLANATION'],
              data.sessionId,
              'TUTORIAL'
            ),
          ],
        },
      },
    ];
  } else if (
    event.input === TUTORIAL_STEPS['PROVIDE_PERMISSION_SETUP_WITH_EXPLANATION']
  ) {
    replies = [
      {
        ...createPermissionSetupDialog(explanPersmissionSetup),
        quickReply: {
          items: [
            createQuickReplyPostbackItem(
              TUTORIAL_STEPS['SETUP_DONE'],
              data.sessionId,
              'TUTORIAL'
            ),
            createQuickReplyPostbackItem(
              TUTORIAL_STEPS['SETUP_LATER'],
              data.sessionId,
              'TUTORIAL'
            ),
          ],
        },
      },
    ];
  } else if (event.input === TUTORIAL_STEPS['SETUP_DONE']) {
    replies = [createEndingMessage()];
  } else if (event.input === TUTORIAL_STEPS['SETUP_LATER']) {
    replies = [
      {
        type: 'text',
        text: replySetupLater,
      },
      createEndingMessage(),
    ];
  } else {
    throw new Error('input undefined');
  }

  const visitor = ga(userId, 'TUTORIAL');
  visitor.event({
    ec: 'Tutorial',
    ea: 'Step',
    el: Object.keys(TUTORIAL_STEPS).find(
      key => TUTORIAL_STEPS[key] === event.input
    ),
  });
  visitor.send();

  return { data, event, issuedAt, userId, replies, isSkipUser };
}
