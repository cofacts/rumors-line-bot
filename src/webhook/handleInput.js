import { t } from 'ttag';
import initState from './handlers/initState';
import defaultState from './handlers/defaultState';
import { ManipulationError } from './handlers/utils';
import { extractArticleId } from 'src/lib/sharedUtils';
import tutorial, { TUTORIAL_STEPS } from './handlers/tutorial';
import handlePostback from './handlePostback';

/**
 * Given input event and context, outputs the new context and the reply to emit.
 *
 * @param {Object<data>} context The current context of the bot
 * @param {*} event The input event
 * @param {*} userId LINE user ID that does the input
 */
export default async function handleInput({ data = {} }, event, userId) {
  let state;
  let replies;
  let isSkipUser = false;

  if (event.input === undefined) {
    throw new Error('input undefined');
  }

  if (event.type === 'message') {
    // Trim input because these may come from other chatbot
    //
    const trimmedInput = event.input.trim();
    const articleId = extractArticleId(trimmedInput);
    if (articleId) {
      // Start new session, reroute to CHOOSING_ARTILCE and simulate "choose article" postback event
      data = {
        // Start a new session
        sessionId: Date.now(),
        searchedText: '',
      };
      event = {
        type: 'postback',
        input: articleId,
      };
      return await handlePostback({ data }, 'CHOOSING_ARTICLE', event, userId);
    } else if (event.input === TUTORIAL_STEPS['RICH_MENU']) {
      state = 'TUTORIAL';
    } else {
      // The user forwarded us an new message.
      // Create a new "search session".
      //
      data = {
        // Used to determine button postbacks and GraphQL requests are from
        // previous sessions
        //
        sessionId: Date.now(),
      };
      state = '__INIT__';
    }
  } else {
    state = 'Error';
  }

  let params = {
    data,
    state,
    event,
    userId,
    replies,
    isSkipUser,
  };

  // Sets data and replies
  //
  do {
    params.isSkipUser = false;
    try {
      switch (params.state) {
        case '__INIT__': {
          params = await initState(params);
          break;
        }
        case 'TUTORIAL': {
          params = tutorial(params);
          break;
        }

        default: {
          params = defaultState(params);
          break;
        }
      }
    } catch (e) {
      if (e instanceof ManipulationError) {
        params = {
          ...params,
          replies: [
            {
              type: 'flex',
              altText: e.toString(),
              contents: {
                type: 'bubble',
                header: {
                  type: 'box',
                  layout: 'vertical',
                  contents: [
                    {
                      type: 'text',
                      text: `⚠️ ${t`Wrong usage`}`,
                      color: '#ffb600',
                      weight: 'bold',
                    },
                  ],
                },
                body: {
                  type: 'box',
                  layout: 'vertical',
                  contents: [
                    {
                      type: 'text',
                      text: e.message,
                      wrap: true,
                    },
                  ],
                },
                styles: {
                  body: {
                    separator: true,
                  },
                },
              },
            },
          ],
        };
      } else {
        throw e;
      }
    }
    ({ isSkipUser } = params);
  } while (isSkipUser);

  ({ data, replies } = params);

  return {
    context: { data },
    replies,
  };
}
