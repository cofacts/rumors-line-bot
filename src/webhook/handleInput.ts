import initState from './handlers/initState';
import defaultState from './handlers/defaultState';
import { extractArticleId } from 'src/lib/sharedUtils';
import { TUTORIAL_STEPS } from './handlers/tutorial';
import handlePostback from './handlePostback';
import {
  ChatbotState,
  ChatbotStateHandlerParams,
  ChatbotStateHandlerReturnType,
} from 'src/types/chatbotState';
import { Result } from 'src/types/result';
import { Message, MessageEvent, TextEventMessage } from '@line/bot-sdk';

/**
 * Given input event and context, outputs the new context and the reply to emit.
 *
 * @param context The current context of the bot
 * @param event The input event
 * @param userId LINE user ID that does the input
 */
export default async function handleInput(
  { data = {} },
  event: MessageEvent & { message: TextEventMessage },
  userId: string
): Promise<Result> {
  let state: ChatbotState;
  const replies: Message[] = [];

  // Trim input because these may come from other chatbot
  //
  const trimmedInput = event.message.text.trim();
  const articleId = extractArticleId(trimmedInput);
  if (articleId) {
    // Start new session, reroute to CHOOSING_ARTILCE and simulate "choose article" postback event
    const sessionId = Date.now();
    return await handlePostback(
      // Start a new session
      {
        sessionId,
        searchedText: '',
      },
      {
        state: 'CHOOSING_ARTICLE',
        sessionId,
        input: articleId,
      },
      userId
    );
  } else if (event.message.text === TUTORIAL_STEPS['RICH_MENU']) {
    // Start new session, reroute to TUTORIAL
    const sessionId = Date.now();
    return await handlePostback(
      { sessionId, searchedText: '' },
      {
        state: 'TUTORIAL',
        sessionId,
        input: TUTORIAL_STEPS['RICH_MENU'],
      },
      userId
    );
  } else {
    // The user forwarded us an new message.
    // Create a new "search session".
    //
    data = {
      // Used to determine button postbacks and GraphQL requests are from
      // previous sessions
      //
      sessionId: Date.now(),
    };
    state = '__INIT__';
  }

  const params: ChatbotStateHandlerParams = {
    data,
    state,
    event: {
      ...event,
      message: event.message,
      input: trimmedInput,
    },
    userId,
    replies,
  };

  let result: ChatbotStateHandlerReturnType;

  // Sets data and replies
  //
  switch (params.state) {
    case '__INIT__': {
      result = await initState(params);
      break;
    }

    default: {
      result = defaultState(params);
      break;
    }
  }

  return {
    context: { data: result.data },
    replies: result.replies,
  };
}
