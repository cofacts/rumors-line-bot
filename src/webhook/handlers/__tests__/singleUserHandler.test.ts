import MockDate from 'mockdate';
import handleInput from '../handleInput';
import originalInitState from '../initState';
import originalHandlePostback from '../handlePostback';
import { TUTORIAL_STEPS } from '../tutorial';

import { VIEW_ARTICLE_PREFIX, getArticleURL } from 'src/lib/sharedUtils';
import { MessageEvent, TextEventMessage } from '@line/bot-sdk';

jest.mock('../handlers/initState');
jest.mock('../handlePostback');

const initState = originalInitState as jest.MockedFunction<
  typeof originalInitState
>;
const handlePostback = originalHandlePostback as jest.MockedFunction<
  typeof originalHandlePostback
>;

// If session is renewed, sessionId will become this value
const NOW = 1561982400000;

beforeEach(() => {
  initState.mockClear();
  handlePostback.mockClear();
  MockDate.set(NOW);
});

afterEach(() => {
  MockDate.reset();
});

function createTextMessageEvent(
  input: string
): MessageEvent & { message: Pick<TextEventMessage, 'type' | 'text'> } {
  return {
    type: 'message',
    message: {
      id: '',
      type: 'text',
      text: input,
    },
    mode: 'active',
    timestamp: 0,
    source: {
      type: 'user',
      userId: '',
    },
    replyToken: '',
  };
}

it('shows reply list when VIEW_ARTICLE_PREFIX is sent', async () => {
  const event = createTextMessageEvent(
    `${VIEW_ARTICLE_PREFIX}${getArticleURL('article-id')}`
  );

  handlePostback.mockImplementationOnce((data) => {
    return Promise.resolve({
      context: { data },
      replies: [],
    });
  });

  await expect(handleInput(event, 'user-id')).resolves.toMatchInlineSnapshot(`
          Object {
            "context": Object {
              "data": Object {
                "searchedText": "",
                "sessionId": 1561982400000,
              },
            },
            "replies": Array [],
          }
        `);

  expect(handlePostback).toHaveBeenCalledTimes(1);
  expect(handlePostback.mock.calls).toMatchInlineSnapshot(`
    Array [
      Array [
        Object {
          "searchedText": "",
          "sessionId": 1561982400000,
        },
        Object {
          "input": "article-id",
          "sessionId": 1561982400000,
          "state": "CHOOSING_ARTICLE",
        },
        "user-id",
      ],
    ]
  `);
});

it('shows reply list when article URL is sent', async () => {
  const event = createTextMessageEvent(
    getArticleURL('article-id') + '  \n  ' /* simulate manual input */
  );

  handlePostback.mockImplementationOnce((data) => {
    return Promise.resolve({
      context: { data },
      replies: [],
    });
  });

  await expect(handleInput(event, 'user-id')).resolves.toMatchInlineSnapshot(`
          Object {
            "context": Object {
              "data": Object {
                "searchedText": "",
                "sessionId": 1561982400000,
              },
            },
            "replies": Array [],
          }
        `);

  expect(handlePostback).toHaveBeenCalledTimes(1);
  expect(handlePostback.mock.calls).toMatchInlineSnapshot(`
    Array [
      Array [
        Object {
          "searchedText": "",
          "sessionId": 1561982400000,
        },
        Object {
          "input": "article-id",
          "sessionId": 1561982400000,
          "state": "CHOOSING_ARTICLE",
        },
        "user-id",
      ],
    ]
  `);
});

it('Resets session on free-form input, triggers fast-forward', async () => {
  const input = 'Newly forwarded message';
  const event = createTextMessageEvent(input);

  initState.mockImplementationOnce(({ data }) => {
    return Promise.resolve({
      data,
      replies: [],
    });
  });

  await expect(handleInput(event, 'user-id')).resolves.toMatchInlineSnapshot(`
          Object {
            "context": Object {
              "data": Object {
                "searchedText": "Newly forwarded message",
                "sessionId": 1561982400000,
              },
            },
            "replies": Array [],
          }
        `);

  expect(initState).toHaveBeenCalledTimes(1);
  expect(initState.mock.calls).toMatchInlineSnapshot(`
    Array [
      Array [
        Object {
          "data": Object {
            "searchedText": "Newly forwarded message",
            "sessionId": 1561982400000,
          },
          "userId": "user-id",
        },
      ],
    ]
  `);
});

describe('tutorial', () => {
  it('handles tutorial trigger from rich menu', async () => {
    const event = createTextMessageEvent(TUTORIAL_STEPS['RICH_MENU']);

    handlePostback.mockImplementationOnce((data) => {
      return Promise.resolve({
        context: { data },
        replies: [],
      });
    });

    await expect(handleInput(event, 'user-id')).resolves.toMatchInlineSnapshot(`
            Object {
              "context": Object {
                "data": Object {
                  "searchedText": "",
                  "sessionId": 1561982400000,
                },
              },
              "replies": Array [],
            }
          `);

    expect(handlePostback).toHaveBeenCalledTimes(1);
  });
});
