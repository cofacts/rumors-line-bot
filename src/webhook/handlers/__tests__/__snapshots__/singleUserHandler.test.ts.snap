// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handles follow and unfollow event: Tutorial replies 1`] = `
Array [
  Array [
    "/message/reply",
    Object {
      "messages": Array [
        Object {
          "altText": "This is a chatbot that looks up suspicious forwarded messages for you. Here is how to use me:",
          "contents": Object {
            "contents": Array [
              Object {
                "body": Object {
                  "contents": Array [
                    Object {
                      "aspectMode": "cover",
                      "aspectRatio": "1:1",
                      "gravity": "top",
                      "size": "full",
                      "type": "image",
                      "url": "/static/img/greeting.png?cachebust=20201105",
                    },
                    Object {
                      "contents": Array [
                        Object {
                          "size": "lg",
                          "text": "This is a chatbot that looks up suspicious forwarded messages for you. Here is how to use me:",
                          "type": "text",
                          "weight": "bold",
                          "wrap": true,
                        },
                      ],
                      "flex": 1,
                      "justifyContent": "center",
                      "layout": "vertical",
                      "offsetBottom": "0px",
                      "offsetEnd": "0px",
                      "offsetStart": "0px",
                      "paddingAll": "20px",
                      "paddingTop": "18px",
                      "position": "relative",
                      "type": "box",
                    },
                  ],
                  "layout": "vertical",
                  "paddingAll": "0px",
                  "type": "box",
                },
                "type": "bubble",
              },
            ],
            "type": "carousel",
          },
          "type": "flex",
        },
        Object {
          "altText": "1. When receiving a message from elsewhere
2. Long press and share
3. Select Cofacts to share
4. Cofacts replies with a crowd-sourced fact-check or chatbot replies",
          "contents": Object {
            "contents": Array [
              Object {
                "body": Object {
                  "contents": Array [
                    Object {
                      "aspectMode": "cover",
                      "aspectRatio": "1:1",
                      "gravity": "top",
                      "size": "full",
                      "type": "image",
                      "url": "/static/img/tutorial1.png?cachebust=20201105",
                    },
                    Object {
                      "contents": Array [
                        Object {
                          "size": "lg",
                          "text": "1. When receiving a message from elsewhere",
                          "type": "text",
                          "weight": "bold",
                          "wrap": true,
                        },
                      ],
                      "flex": 1,
                      "justifyContent": "center",
                      "layout": "vertical",
                      "offsetBottom": "0px",
                      "offsetEnd": "0px",
                      "offsetStart": "0px",
                      "paddingAll": "20px",
                      "paddingTop": "18px",
                      "position": "relative",
                      "type": "box",
                    },
                  ],
                  "layout": "vertical",
                  "paddingAll": "0px",
                  "type": "box",
                },
                "type": "bubble",
              },
              Object {
                "body": Object {
                  "contents": Array [
                    Object {
                      "aspectMode": "cover",
                      "aspectRatio": "1:1",
                      "gravity": "top",
                      "size": "full",
                      "type": "image",
                      "url": "/static/img/tutorial2.png?cachebust=20201105",
                    },
                    Object {
                      "contents": Array [
                        Object {
                          "size": "lg",
                          "text": "2. Long press and share",
                          "type": "text",
                          "weight": "bold",
                          "wrap": true,
                        },
                      ],
                      "flex": 1,
                      "justifyContent": "center",
                      "layout": "vertical",
                      "offsetBottom": "0px",
                      "offsetEnd": "0px",
                      "offsetStart": "0px",
                      "paddingAll": "20px",
                      "paddingTop": "18px",
                      "position": "relative",
                      "type": "box",
                    },
                  ],
                  "layout": "vertical",
                  "paddingAll": "0px",
                  "type": "box",
                },
                "type": "bubble",
              },
              Object {
                "body": Object {
                  "contents": Array [
                    Object {
                      "aspectMode": "cover",
                      "aspectRatio": "1:1",
                      "gravity": "top",
                      "size": "full",
                      "type": "image",
                      "url": "/static/img/tutorial3.png?cachebust=20201105",
                    },
                    Object {
                      "contents": Array [
                        Object {
                          "size": "lg",
                          "text": "3. Select Cofacts to share",
                          "type": "text",
                          "weight": "bold",
                          "wrap": true,
                        },
                      ],
                      "flex": 1,
                      "justifyContent": "center",
                      "layout": "vertical",
                      "offsetBottom": "0px",
                      "offsetEnd": "0px",
                      "offsetStart": "0px",
                      "paddingAll": "20px",
                      "paddingTop": "18px",
                      "position": "relative",
                      "type": "box",
                    },
                  ],
                  "layout": "vertical",
                  "paddingAll": "0px",
                  "type": "box",
                },
                "type": "bubble",
              },
              Object {
                "body": Object {
                  "contents": Array [
                    Object {
                      "aspectMode": "cover",
                      "aspectRatio": "1:1",
                      "gravity": "top",
                      "size": "full",
                      "type": "image",
                      "url": "/static/img/tutorial4.png?cachebust=20201105",
                    },
                    Object {
                      "contents": Array [
                        Object {
                          "size": "lg",
                          "text": "4. Cofacts replies with a crowd-sourced fact-check or chatbot replies",
                          "type": "text",
                          "weight": "bold",
                          "wrap": true,
                        },
                      ],
                      "flex": 1,
                      "justifyContent": "center",
                      "layout": "vertical",
                      "offsetBottom": "0px",
                      "offsetEnd": "0px",
                      "offsetStart": "0px",
                      "paddingAll": "20px",
                      "paddingTop": "18px",
                      "position": "relative",
                      "type": "box",
                    },
                  ],
                  "layout": "vertical",
                  "paddingAll": "0px",
                  "type": "box",
                },
                "type": "bubble",
              },
              Object {
                "body": Object {
                  "contents": Array [
                    Object {
                      "size": "lg",
                      "text": "Wanna try it out? Just forward a message to me!",
                      "type": "text",
                      "weight": "bold",
                      "wrap": true,
                    },
                  ],
                  "layout": "vertical",
                  "type": "box",
                },
                "footer": Object {
                  "contents": Array [
                    Object {
                      "action": Object {
                        "data": "{\\"input\\":\\"Simulates forwarding a message\\",\\"sessionId\\":1561982400000,\\"state\\":\\"TUTORIAL\\"}",
                        "displayText": "只要每天早上一杯地瓜葉牛奶。不僅有效降低三高，甚至連痛風也沒了；此外，地瓜葉牛奶的作法也很簡單，只要先將地瓜葉川燙過後，再加入鮮奶打成汁即可。",
                        "label": "Simulates forwarding a message",
                        "type": "postback",
                      },
                      "color": "#ffb600",
                      "style": "primary",
                      "type": "button",
                    },
                  ],
                  "layout": "horizontal",
                  "type": "box",
                },
                "type": "bubble",
              },
            ],
            "type": "carousel",
          },
          "type": "flex",
        },
      ],
      "replyToken": "nHuyWiB7yP5Zw52FIkcQobQuGDXCTA",
    },
  ],
]
`;

exports[`handles follow and unfollow event: User settings should have notification turned on 1`] = `Array []`;
